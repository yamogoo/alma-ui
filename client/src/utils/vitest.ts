import type { VueWrapper } from "@vue/test-utils";

export const getTypedEmittedEvent = <V>(
  wrapper: VueWrapper,
  eventName: string,
) => {
  return wrapper.emitted(eventName) as V[][];
};
