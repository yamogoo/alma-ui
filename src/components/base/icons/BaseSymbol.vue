<script setup lang="ts">
import { defineAsyncComponent, computed } from "vue";

import type {
  SymbolName,
  SymbolStyle,
  SymbolWeight,
  SymbolColor,
  SymbolSize,
} from "./symbols";
import BaseSkeleton from "@/components/base/skeleton/BaseSkeleton.vue";

const props = defineProps<Props>();

const symbol = computed(() => {
  const name = props.name;
  const style = props.style;
  const weight = props.weight;

  return defineAsyncComponent(async () => {
    return import(`../../../assets/icons/${name}_${style}_${weight}.svg`);
  });
});
</script>

<script lang="ts">
export interface Props {
  name: SymbolName;
  style: SymbolStyle;
  weight: SymbolWeight;
  color?: SymbolColor;
  size?: SymbolSize;
}
</script>

<template>
  <div class="icon" data-testid="icon" :class="[size, color]">
    <Suspense>
      <component v-if="symbol" :is="symbol" viewBox="0 0 24 24"></component>
      <template #fallback>
        <BaseSkeleton></BaseSkeleton>
      </template>
    </Suspense>
  </div>
</template>

<style lang="scss">
@use "sass:map";

@mixin defineIconSizes($map: $icon) {
  @each $size, $val in $map {
    &.#{$size} {
      @include box(px2rem(map.get($val, "size")));
    }
  }
}

.icon {
  @include box(auto, inherit);
  line-height: 0;

  svg {
    @include box(auto, inherit);
    fill: inherit;

    path {
      fill: inherit;
    }
  }

  .skeleton {
    @include box(100%);
  }

  /* * * Sizes * * */

  @include defineIconSizes();

  /* * * Colors * * */

  &.primary {
    @include themify($themes) {
      fill: themed("label", "primary");
    }
  }

  &.primary-inversed {
    @include themify($themes) {
      fill: themed("label", "primary-inversed");
    }
  }

  &.secondary {
    @include themify($themes) {
      fill: themed("label", "secondary");
    }
  }

  &.secondary-inversed {
    @include themify($themes) {
      fill: themed("label", "secondary-inversed");
    }
  }

  &.disabled {
    @include themify($themes) {
      fill: themed("label", "disabled");
    }
  }

  &.accent {
    @include themify($themes) {
      fill: themed("label", "accent");
    }
  }
}
</style>
