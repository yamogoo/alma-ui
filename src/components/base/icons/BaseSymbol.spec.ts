import { describe, expect, test, vi } from "vitest";
import { mount } from "@vue/test-utils";

import type { SymbolName, SymbolStyle, SymbolWeight } from "./symbols";

import BaseSymbol from "./BaseSymbol.vue";

const icons: Array<SymbolName> = ["back", "check"];

const weightList: Array<SymbolWeight> = ["100", "200", "300", "400", "500"];

describe("BaseSymbol", () => {
  test.each(icons)("should render %s icon from the Icons Set", async (name) => {
    const _name = name as SymbolName;
    const _style: SymbolStyle = "outline";

    for (const weight of weightList) {
      const wrapper = mount(BaseSymbol, {
        props: {
          name: _name,
          weight: weight,
          style: _style,
        },
      });

      await vi.dynamicImportSettled();

      const res = wrapper.find("svg");

      expect(res.exists()).toBe(true);
      expect(res.exists()).toMatchSnapshot();
    }
  });
});
