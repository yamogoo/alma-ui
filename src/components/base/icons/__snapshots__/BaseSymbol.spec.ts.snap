// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BaseSymbol > should render back icon from the Icons Set 1`] = `true`;

exports[`BaseSymbol > should render back icon from the Icons Set 2`] = `true`;

exports[`BaseSymbol > should render back icon from the Icons Set 3`] = `true`;

exports[`BaseSymbol > should render back icon from the Icons Set 4`] = `true`;

exports[`BaseSymbol > should render back icon from the Icons Set 5`] = `true`;

exports[`BaseSymbol > should render check icon from the Icons Set 1`] = `true`;

exports[`BaseSymbol > should render check icon from the Icons Set 2`] = `true`;

exports[`BaseSymbol > should render check icon from the Icons Set 3`] = `true`;

exports[`BaseSymbol > should render check icon from the Icons Set 4`] = `true`;

exports[`BaseSymbol > should render check icon from the Icons Set 5`] = `true`;
